---
title: "GWAS SurvivR Vignette"
author:
- name: Abbas Rizvi
  affiliation: The Ohio State University, Columbus, OH
- name: Ezgi Karaesmen
  affiliation: The Ohio State University, Columbus, OH
- name: Martin Morgan
  affiliation: Roswell Park Cancer Institute, Buffalo, NY    
- name: Lara Sucheston-Campbell  
  affiliation: The Ohio State University, Columbus, OH
package: GWAS SurvivR
date: "November 13, 2017"
output:
        BiocStyle::html_document2
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}          
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(tidyverse)
library(SurvivR)
library(magrittr)
library(data.table)
library(SummarizedExperiment)
library(GWASTools)
```

# Introduction
This package provides a workflow for how to use imputed GWAS data from either IMPUTE2 or the Michigan server and conduct survival analysis. Survival analyses are methods for analyzing time-to-event outcomes. This vignette desribes the data structures that are used in these analyses and provides a tutorial using simulated data consisting of 10,000 SNPs and 1,500 samples. In a normal GWAS, users would impute genetic data from typically a chip of genotyping 750,000-1 million SNPs. The data would either be partitioned (e.g. 5Mb chromosome chunks until end of chromosome) or full chromosome.  

# Getting started
Install GWAS SurvivR from the [Sucheston-Campbell Lab Github repository](http://github.com/suchestoncampbelllab/survivR) using `devtools`.

```{r, eval=F}
devtools::install_github("suchestoncampbelllab/survivR")
library(SurvivR)
```

**Note:** This package depends on `GWASTools` which uses netcdf framework on linux. Therefore, for linux users, please install `libnetcdf-dev` and `netcdf-bin` before installing `SurvivR`. 

# Basics of survival analysis in GWAS setting
To test for association using GWAS data, often times investigators are interested in determining genetic associations to surivival. To perform such analyses, phenotypic or covariate data must be available that denote some event and time (occurrences) of said events. The most common implementation of these analyses is the Cox proportional hazard model. Th

## Dosage format
Genotype probabilities are converted into dosage format. Let $gp_{00}$, $gp_{01}$, and $gp_{11}$ denote the probability that a sample has 0 alleles coded 1, 1 allele coded 1, and 2 alleles coded 1 respectively. We can convert these three values into a single numeric value called dosage. 

$$d = gp_{01} + 2gp_{11}$$
SNP dosages are then added to the Cox model for survival analysis. 


# IMPUTE2 data
## File structure
IMPUTE2 outputs 3 files required for analysis:   

  - Genotype file (`.impute`)  
  - Sample file (`.sample`)  
  - Info file (`.info`)    
  
[More information can be read about these formats](http://www.stats.ox.ac.uk/~marchini/software/gwas/file_format.html)
 
```{r, echo=F}
# 1500 patients
cols.geno <- c("snpid", "rsid", "position", "allele0", "allele1", paste0(paste0("sample",rep(1:1500, each=3)), "_", rep(c("gp00", "gp01", "gp11"), 600)))
example.impute <- fread("~/Google Drive/OSU_PHD/survivR/data/example.impute")
colnames(example.impute) <- cols.geno
example.sample <- fread("~/Google Drive/OSU_PHD/survivR/data/example.sample")
example.info <- fread("../data/example.info")
```
 
### Genotype file format
 
```{r, echo=F}
cols <- c("snpid",
          "rsid",
          "position",
          "allele0",
          "allele1",
          "sample1_gp00",
          "sample1_gp01",
          "sample1_gp11",
          "sample2_gp00",
          "sample2_gp01",
          "sample2_gp11",
          "...",
          "sampleN_gp00",
          "sampleN_gp01",
          "sampleN_gp11")
descrip <- c("genotyped (refseq id of SNP) or imputed (- - -)",
             "refseq id of SNP",
             "physical genomic location (base pair)",
             "typically major allele",
             "typically minor allele",
             "sample1 genotype probability of 2 copies of allele0",
             "sample1 genotype probability of 1 copy allele0, 1 copy allele1",
             "sample1 genotype probability of 2 copies of allele1",
             "sample2 genotype probability of 2 copies of allele0",
             "sample2 genotype probability of 1 copy allele0, 1 copy allele1",
             "sample2 genotype probability of 2 copies of allele1",
             "...",
             "sampleN genotype probability of 2 copies of allele0",
             "sampleN genotype probability of 1 copy allele0, 1 copy allele1",
             "sampleN genotype probability of 2 copies of allele1")
kable(data.frame(`Column Name`=cols, "Description"=descrip), caption="Genotype Format Column Description")
```

```{r}
example.impute[1:5,1:11]
```


### Sample file format
The sample file format has three parts:

1. Header line:  
  - This line has a minimum of three entries which are `ID_1`, `ID_2`, and `missing`. Any columns beyond the first 3 are additional columns. 


```{r, echo=F}
col.samps <- c("ID_1", "ID_2", "missing", "cov_1", "...", "cov_n", "pheno1", "bin1")
descrip.samps <- c("first ID",
                   "second ID",
                   "missing data proportion for each sample (-9 indicates missing data)",
                   "covariate 1",
                   "...",
                   "Nth covariate",
                   "continuous phenotype",
                   "binary phenotype")
samp <- data.frame("Column Names"=col.samps, "Description"=descrip.samps)
kable(samp, caption="Sample File Column Descriptions")
```
  
2. Line detailing how the types of variables are stored
  - The first three entries are set to `0`. Subsequent entries in this row describe the covariates and phenotypes. 

```{r, echo=F}
samp.l2 <- c("D", "C", "P", "B")
samp.l2.desc <- c("Discrete covariate", "Continuous covariate", "Continuous Phenotype", "Binary Phenotypes (0=Controls, 1=Cases)")
kable(data.frame("Variables"=samp.l2, "Description"=samp.l2.desc), caption="Second row descriptors in sample file")
```

3. Individual information
- The remainder of the `.sample` file consists of lines containing information for each of the columns headers. 

**Note:** Missing values are set to `-9` by default. There is some history behind this value and thats why it keeps getting used.

For our example, we will only have a discrete variable `sex` as our additional column.

```{r}
head(example.sample)
```


### Info file format  
- The [info file from IMPUTE2 description](https://mathgen.stats.ox.ac.uk/impute/output_file_options.html) is as follows:  
```{r, echo=F}
info.cols <- c("snpid", "rsid", "position", "exp_freq_a1", "info", "certainty", "type")
info.descrip <- c("ref seq identifier or imputed annotation (- - -)" ,
                  "ref seq identifier",
                  "base pair position",
                  "expected frequency of allele coded 1",
                  "measure of observed statistical information with the allele frequency estimate",
                  "average certainty of best guess-genotypes",
                  "internal type assigned to SNP")
kable(data.frame("Column Name"=info.cols, "Description"=info.descrip), caption="Info file header description")
```

Each row corresponds to a typed or imputed SNP and describes some quality measures of that SNP.

```{r}
head(example.info)
```

## Convert IMPUTE2 data to GDS
Due to the size of genotype data, which in a typical GWAS can be quite large (full chromosome ~20GB, chromosome chunks ~1.5GB), we first convert the genotype data into [genomic data structure (GDS) format](https://github.com/zhengxwen/gdsfmt) by re-implementing a function from the package `GWASTools` which we call `convertImputeGds`. This will compress the imputed data substantially, as well as convert the data into dosage format. 


`convertImputeGds` has 
To convert the IMPUTE2 data to GDS format, we need to invoke our function `convertImputeGds`. This function takes 4 arguments:  

1. chunk.impute -- a character string with the specific name of `.impute` file
2. chunk.sample -- a character string with the specific name of `.sample` file 
3. chr -- a numeric vector with the chromosome being analyzed, this has to be input, regardless if there are multiple chromosomes in the imputed file.
4. outfile.name -- a one element character string with the name of the output files after conversion (do not add extensions)  

```{r, warning=F, cache=T}
library(SurvivR)
convertImputeGds("~/Google Drive/OSU_PHD/survivR/data/example.impute",
                 "~/Google Drive/OSU_PHD/survivR/data/example.sample",
                 2,
                 "~/Google Drive/OSU_PHD/survivR/data/example")
```
This function creates 3 files that will be saved to your working directory (or an otherwise user specific path). The files are stored in your specified path or working directory and named `example.gds`, `example.snp.rdata`, and `example.scan.info`. 

## Read in files as SummarizedExperiment
Since users may be interested in seeing the data within an R/Bioconductor framework, we read in our GDS files are `SummarizedExperiment` objects to allow for access to 

```{r}
se <- readImputedGds(gdsfile = "~/Google Drive/OSU_PHD/survivR/data/example.gds",
                     scanfile = "~/Google Drive/OSU_PHD/survivR/data/example.scan.rdata",
                     snpfile = "~/Google Drive/OSU_PHD/survivR/data/example.snp.rdata",
                     infofile = "~/Google Drive/OSU_PHD/survivR/data/example.info")
se
assay(se)[1:5,1:10]
rowRanges(se)
colData(se)
```
## Flip dosage
Users can flip the dosage if needed.

## Add covariates
Covariates can be added to the `colData` slot of the `SummarizedExperiment` by using the `addCov` function. It is required that the first column is the sample ID. These sample ID must match the IDs that are in the `ID_2` column of `colData(se)`.

```{r, echo=F}
cov.cols <- c("ID_2", "event", "time", "age", "bmiOVWT", "sex")
cov.ids <- colData(se)$ID_2
event <-  sample(seq(0,1), 1500, replace=T)
time <-  sample(seq(0,12,by=0.01), 1500, replace=T)
age <- sample(seq(18,65,by=0.001), 1500, replace=T)
bmiOVWT <- sample(seq(0,1), 1500, replace=T, prob=c(0.25, 0.75))
sex <- sample(seq(0,1), 1500, replace=T, prob=c(0.35, 0.65))
covfile <- data.frame(cbind(cov.ids, event, time, age, bmiOVWT,sex))
colnames(covfile) <- cov.cols
#write.table(covfile, file="example.covariate", sep="\t", row.names=F, quote=F)
```

```{r}
covfile <- "example.covariate"
se <- addCov(se, covfile)
```

To check and see if the covariate merger was successful:

```{r}
colData(se)
```

## Run survival
Now we can run survival analysis using CoxPH


# Michigan Imputation Server Output

## File structure

### Compressed VCF format

## Run survival


# Session info
Here is the output of `sessionInfo()` on the system that this document was compiled:
```{r, echo=F}
sessionInfo()
```
